compressors/shannon_coder.py

```diff
@@ -0,0 +1,178 @@
+"""Shannon Coding
+Shannon coding is a basic prefix free code. There is some confusion in literature about Shannon, Fano and
+Shannon-Fano codes, e.g. see Naming section of https://en.wikipedia.org/wiki/Shannon–Fano_coding.
+
+This document uses cumulative probability based method for shannon coding as described in the wiki article above.
+"""
+from dataclasses import dataclass
+from typing import Any, Tuple
+from utils.bitarray_utils import float_to_bitarrays, BitArray
+from utils.test_utils import get_random_data_block, try_lossless_compression
+from compressors.prefix_free_compressors import PrefixFreeEncoder, PrefixFreeDecoder, PrefixFreeTree, BinaryNode
+from core.prob_dist import ProbabilityDist
+import math
+
+
+@dataclass
+class ShannonNode(BinaryNode):
+    """represents a node of the Shannon Tree
+    Adds a prefix of code to BinaryNode which provides left_child, right_child and id to allow making PrefixFreeTree
+    from code.
+    """
+    code: BitArray = None
+
+
+class ShannonTree(PrefixFreeTree):
+    """
+    Generates Codewords based on Shannon Coding. Allows generating PrefixFreeTree from the obtained codewords.
+    """
+
+    def __init__(self, prob_dist: ProbabilityDist):
+        self.prob_dist = prob_dist
+        # sort the probability distribution and get cumulative probability which will be used for encoding
+        self.sorted_prob_dist = ProbabilityDist.get_sorted_prob_dist(prob_dist.prob_dict)
+        self.cum_prob_dict = self.sorted_prob_dist.cumulative_prob_dict
+        # construct the tree and set the root_node of PrefixFreeTree base class
+        super().__init__(root_node=self.build_shannon_tree())
+
+    def encode_alphabet(self, symbol) -> BitArray:
+        # compute the mid-point corresponding to the range of the given symbol
+        cum_prob = self.cum_prob_dict[symbol]
+
+        # compute encode length
+        encode_len = math.ceil(self.prob_dist.log_probability(symbol))
+
+        # the encode is the binary representation of the cumulative probability in ascending order,
+        _, code = float_to_bitarrays(cum_prob, encode_len)
+        return code
+
+    @staticmethod
+    def build_tree_from_code(symbol, code, root_node) -> BinaryNode:
+        """ function to generate prefix-free tree from code.
+        Args:
+            symbol: current symbol
+            code: current code
+            root_node: root node to the ShannonTree
+
+        Returns:
+            the pointer to root node of the tree so far
+        """
+        code_so_far = BitArray()
```

Comment:
A couple of suggestions here: 
1. Try to reduce code repetition, as that makes code longer + can lead to bugs. 
2. Add inline comments while you code, so that you don't need to add them later + it might help your thought process. 

Also, here is a suggestion for changing you code.. you might have to get it working, as I have not tested this: 
```python
# initialize the curr_node, code_so_far temporary var
curr_node = self.root_node
code_so_far = BitArray()

for i, bit in enumerate(code):
    code_so_far.append(bit)

    # get a pointer to the child node 
    child = curr_node.right_child if bit else curr_node.left_child

    if child is None:
        # add a new node, in case it is not already present
        new_node = ...
        child = new_node
    curr_node = child
```

Reply:
@tpulkit: re-pasting my comment, as that one became outdated